src/main/java/org/terasology/gooeysQuests/quests/mines/MinesQuestSystem.java

```diff
@@ -0,0 +1,202 @@
+    private Random random = new Random();
+    private Predicate<Block> isAirCondition;
+    private Predicate<Block> isGroundCondition;
+
+    @Override
+    public void postBegin() {
+        isAirCondition = blockPredicateProvider.getBlockPredicate("StructureTemplates:IsAirLike");
+        isGroundCondition = blockPredicateProvider.getBlockPredicate("StructureTemplates:IsGroundLike");
+    }
+
+
+    @ReceiveEvent
+    public void onCreateStartQuestsEvent(CreateStartQuestsEvent event, EntityRef character,
+                                         PersonalQuestsComponent questsComponent) {
+        Prefab questPrefab = assetManager.getAsset("GooeysQuests:MinesQuest", Prefab.class).get();
+        Logger logger = LoggerFactory.getLogger(MinesQuestSystem.class);
+        logger.info("questPrefab: "+questPrefab);
+        EntityBuilder questEntityBuilder = entityManager.newBuilder(questPrefab);
+        questEntityBuilder.setOwner(character);
+        EntityRef entity = questEntityBuilder.build();
+        questsComponent.questsInPreperation.add(entity);
+        character.saveComponent(questsComponent);
+    }
+
+    @ReceiveEvent(components = MinesQuestComponent.class)
+    public void onPrepareQuest(PrepareQuestEvent event, EntityRef quest) {
+        Logger logger = LoggerFactory.getLogger(MinesQuestSystem.class);
+        logger.info("preparing quest");
+        EntityRef owner = quest.getOwner();
+        LocationComponent questOwnerLocation = owner.getComponent(LocationComponent.class);
+        Vector3i questOwnerBlockPos = new Vector3i(questOwnerLocation.getWorldPosition());
+        Vector3i randomPosition = new Vector3i(questOwnerBlockPos);
+        randomPosition.addX(randomHorizontalOffset());
+        randomPosition.addZ(randomHorizontalOffset());
+        Vector3i surfaceGroundBlockPosition = findSurfaceGroundBlockPosition(randomPosition);
+        if (surfaceGroundBlockPosition == null) {
+            logger.info("surface null");
+            return;
+        }
+        EntityRef entranceSpawner = structureTemplateProvider.
+                getRandomTemplateOfType("GooeysQuests:minesEntrance");
+        logger.info("entrance obtained");
+        BlockRegionTransform foundSpawnTransformation = findGoodSpawnTransformation(surfaceGroundBlockPosition,
+                entranceSpawner);
+        if (foundSpawnTransformation == null) {
+            logger.info("no spawn place found");
+            return;
+        }
+
+        questToFoundSpawnPossibilityMap.put(quest, new MinesQuestSystem.FoundSpawnPossibility(entranceSpawner, foundSpawnTransformation));
+        quest.send(new QuestReadyEvent());
+    }
+    private static class FoundSpawnPossibility {
+        private EntityRef entranceSpawner;
+        private BlockRegionTransform transformation;
+
+        public FoundSpawnPossibility(EntityRef entranceSpawner, BlockRegionTransform transformation) {
```

Comment:
IntelliJ tells me this `public` is redundant - probably because it is a private inner class :-)

There are also some line ordering issues in here. Make sure you set up the Checkstyle plugin for your IDE to get these and other warnings